tests/samplers_tests/test_qmc.py

```diff
@@ -246,6 +246,8 @@ def test_sample_relative_sobol() -> None:
     numpy.testing.assert_allclose(samples, ref_samples, rtol=1e-6)
 
 
+# TODO(kstoneriv3): Need to add this test.
```

Comment:
This test was not failing several months ago but now failing. I am wondering if there is a relation of between this and the deprecation of `n_jobs` (#2560). From the error of the test, the error seems to be that sample_id ([code](https://github.com/optuna/optuna/pull/2423/files#diff-f6e75a1153a00e6980c0601537c46273706ea852e87df915ccc913268d1e8063R336)) or cached qmc engine ([code](https://github.com/optuna/optuna/pull/2423/files#diff-f6e75a1153a00e6980c0601537c46273706ea852e87df915ccc913268d1e8063R303)) not being synchronized well among workers.

Reply:
@kstoneriv3 
> Is the sampler shared among workers in the multithread execution (e.g. n_jobs=3)?

Yes, it is. Multi-threaded execution differs from multi-process execution in that memory space is shared among workers. Therefore, the sampler is shared by all threads unless you explicitly copy it for each thread. Why we implement like this is that most samplers in Optuna do not have state across trials, so there is no fatal problem if the same sampler is used by all threads.

In Optuna, it is implicit to store information in storage when you want to make it persistent across trials. For example, how about storing it in study's `system_attrs`?

If I have misunderstood something, please let me know.

Reply:
@HideakiImamura Thank you for the clarification!

Then, the issue here is that `scipy.stats.qmc.QMCEngine` is cached within `QMCSampler`. As the `QMCEngine.fastforward()` is called in an unintended order, some of the `sample_id` is skipped. Here, I assume we have two options that we can take to make `QMCSampler` to work. 

1) Just not cache `QMCEngine`. This will make some overhead for sampling QMC sample especially for large `sample_id` (and therefore, for large `trail_number`)

2) Use thread lock for accessing `QMCEngine`. I am not sure how fast/slow the thread lock is and even not sure how easy it is. One potential downside is that `QMCSampler` is not stateless unlike other samplers, which might be a technical debt in the future.

> In Optuna, it is implicit to store information in storage when you want to make it persistent across trials. For example, how about storing it in study's system_attrs?

Sorry I did not fully understand this line. What did you mean to store in study's `system_attrs`? I assume the `sample_id` is stored there already. 

By the way, I assume that there is no need for reseeding at each thread at https://github.com/optuna/optuna/blob/master/optuna/study/_optimize.py#L118.
Is there any reason why it is here?

Reply:
@kstoneriv3 Thank you for the explanation.

> This will make some overhead for sampling QMC sample especially for large sample_id (and therefore, for large trail_number).

Why are the effects of those overheads so severe for large `sample_id`?

---

> What did you mean to store in study's system_attrs? 

I mean that how about generating the actual samples in the first trial and storing them in the study's system_attrs. In each call of sample_relative, you pop a sample from the system_attrs. But in any case, we need to achieve thread-safe access to the stored samples.

How about using `queue.Queue`? Storing some IDs in the queue, and popping the ID in each call of sample_relative. As far as I know, Python's `queue.Queue` is thread-safe.

---

> Is there any reason why it is here?

I'm not sure but it might be due to historical circumstances. I will ask other team members the reason.

Reply:
> > This will make some overhead for sampling QMC sample especially for large sample_id (and therefore, for large trail_number).
> 
> Why are the effects of those overheads so severe for large `sample_id`?

I benchmarked the overhead for sampling overhead of `QMCEngine` in cases where we instantiate new `QMCEngine` every time and call `QMCEngine.fast_forward(sample_id)` to sample the target sample. Contrary to my expectation, the overhead does not seem to be big (<0.1 sec in most cases). 

For Sobol sequence, I do not see the point in caching the engine. For Halton sequence, one might want to cache it if s/he want the sampling to be very efficient. 

I would not mind the overhead up to 0.1 sec for hyperparameter optimization, so just not caching the `QMCEngine` seems to be a reasonable option.

```python
>>> import time                                                              
>>> from scipy.stats.qmc import Halton, Sobol                                
>>>                                                                          
>>> def time_qmc(sample_id, dim, qmc_type="Halton", rep = 10):               
...     QMCEngine = {                                                        
...         "Halton": scipy.stats.qmc.Halton,                                
...         "Sobol": scipy.stats.qmc.Sobol,                                  
...     }[qmc_type]                                                          
...                                                                          
...     s = time.time()                                                      
...     for i in range(rep):                                                 
...         en = QMCEngine(dim)                                              
...         en.fast_forward(sample_id)                                       
...         en.random(1)                                                     
...     t = time.time()                                                      
...     return (t - s) / rep                                                 
...                                                                          
>>> time_qmc(sample_id=0, dim=100, qmc_type="Halton")                        
0.027872896194458006                                                         
>>> time_qmc(sample_id=1, dim=100, qmc_type="Halton")                        
0.028704166412353516                                                         
>>> time_qmc(sample_id=10, dim=100, qmc_type="Halton")                       
0.02832076549530029                                                          
>>> time_qmc(sample_id=100, dim=100, qmc_type="Halton")                      
0.029372334480285645                                                         
>>> time_qmc(sample_id=1000, dim=100, qmc_type="Halton")                     
0.03932945728302002                                                          
>>> time_qmc(sample_id=10000, dim=100, qmc_type="Halton")                    
0.1251124143600464                                                           
>>>                                                                          
>>> time_qmc(sample_id=0, dim=100, qmc_type="Sobol")
0.0054619073867797855                                                        
>>> time_qmc(sample_id=1, dim=100, qmc_type="Sobol")
0.0054483175277709964                                                        
>>> time_qmc(sample_id=10, dim=100, qmc_type="Sobol")
0.005433773994445801                                                         
>>> time_qmc(sample_id=100, dim=100, qmc_type="Sobol")
0.005469512939453125                                                         
>>> time_qmc(sample_id=1000, dim=100, qmc_type="Sobol")
0.0055207014083862305                                                        
>>> time_qmc(sample_id=10000, dim=100, qmc_type="Sobol")
0.006371212005615234                                                         
>>> 
```

![Figure](https://user-images.githubusercontent.com/30857855/140886302-54e3ccbd-0fcd-428a-8513-6e7f7e79c8f4.png)

<details>
  <summary>code for plotting</summary>

```python
import time
from scipy.stats.qmc import Halton, Sobol
from matplotlib import pyplot

def time_qmc(sample_id, dim, qmc_type="Halton", rep = 10):
    QMCEngine = {
        "Halton": scipy.stats.qmc.Halton,
        "Sobol": scipy.stats.qmc.Sobol,
    }[qmc_type]
    
    s = time.time()
    for i in range(rep):
        en = QMCEngine(dim)
        en.fast_forward(sample_id)
        en.random(1)
    t = time.time()
    return (t - s) / rep


sample_ids = [10**i for i in range(5)]

ave_times_Halton10 = [time_qmc(sample_id=i, dim=10, qmc_type="Halton") for i in sample_ids]
ave_times_Halton100 = [time_qmc(sample_id=i, dim=100, qmc_type="Halton") for i in sample_ids]
ave_times_Sobol10 = [time_qmc(sample_id=i, dim=10, qmc_type="Sobol") for i in sample_ids]
ave_times_Sobol100 = [time_qmc(sample_id=i, dim=100, qmc_type="Sobol") for i in sample_ids]

pyplot.plot(sample_ids, ave_times_Halton10)
pyplot.plot(sample_ids, ave_times_Halton100)
pyplot.plot(sample_ids, ave_times_Sobol10)
pyplot.plot(sample_ids, ave_times_Sobol100)

pyplot.xscale("log")
pyplot.yscale("log")
pyplot.xlabel("sample_id")
pyplot.ylabel("Sampling Overhead per Sample (sec)")
pyplot.legend(["Halton (10 dim)", "Halton (100 dim)", "Sobol (10 dim)", "Halton (100 dim)"])
pyplot.show()
```

</details>

Reply:
@kstoneriv3 I don't know where your benchmark went, just wanted to add some comments.

* You could save a bit on the object instantiation if you call `reset()` instead of creating a new object.
* We made some improvements on `Halton` and in 1.8.0 [`Halton.random`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.qmc.Halton.random.html#scipy.stats.qmc.Halton.random) will handle a `workers` keyword
* When timing small things, prefer using `time.perf_counter` or `timeit`